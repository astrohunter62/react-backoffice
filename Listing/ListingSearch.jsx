import React from 'react'
import PropTypes from 'prop-types'

import ListingSearchBranch from './ListingSearchBranch'

const withListing = Component => class extends React.Component {
  static propTypes = {
    onFilter: PropTypes.func.isRequired,
  }

  constructor() {
    super()

    this.handleClick = this.handleClick.bind(this)
    this.handleFilter = this.handleFilter.bind(this)
  }

  state = {
    open: false,
  }

  handleClick() {
    this.setState({
      open: !this.state.open,
    })

    this.props.onFilter(undefined)
  }

  handleFilter(event) {
    const { value } = event.target

    if (this.state.open) {
      this.props.onFilter(value)
    } else {
      this.props.onFilter(undefined)
    }
  }

  render() {
    return (
      <Component
        {...this.props}
        {...this.state}
        onClick={this.handleClick}
        onFilter={this.handleFilter}
      />
    )
  }
}

export default withListing(ListingSearchBranch)
