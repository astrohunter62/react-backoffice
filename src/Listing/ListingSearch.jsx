import React from 'react'
import PropTypes from 'prop-types'

import ListingSearchBranch from './ListingSearchBranch'

const withListing = Component => class extends React.Component {
  static propTypes = {
    onFilter: PropTypes.func.isRequired,
  }

  constructor() {
    super()

    this.handleClick = this.handleClick.bind(this)
    this.handleFilter = this.handleFilter.bind(this)
    this.getSearchRef = this.getSearchRef.bind(this)
  }

  state = {
    open: false,
  }

  getSearchRef(node) {
    this.searchRef = node
  }

  handleClick() {
    const { open } = this.state
    const { onFilter } = this.props
    const newOpen = !open

    if (this.searchRef && newOpen) {
      this.searchRef.focus()
    }

    this.setState({
      open: newOpen,
    })

    onFilter(undefined)
  }

  handleFilter(event) {
    const { open } = this.state
    const { onFilter } = this.props
    const { value } = event.target

    if (open) {
      onFilter(value)
    } else {
      onFilter(undefined)
    }
  }

  render() {
    return (
      <Component
        {...this.props}
        {...this.state}
        onClick={this.handleClick}
        onFilter={this.handleFilter}
        getSearchRef={this.getSearchRef}
      />
    )
  }
}

export default withListing(ListingSearchBranch)
