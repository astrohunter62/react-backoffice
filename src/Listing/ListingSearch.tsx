import React, { FunctionComponent, useEffect, useRef, useState } from "react";
import classNames from "classnames";
import {
  TextField,
  Tooltip,
  IconButton,
  makeStyles,
  Theme,
} from "@material-ui/core";
import SearchIcon from "@material-ui/icons/Search";

const useStyles = makeStyles((theme: Theme) => ({
  root: {
    position: "relative",
    marginRight: theme.spacing(),
  },
  field: {
    position: "absolute",
    right: "100%",
    width: 0,
    marginTop: theme.spacing(),
    transition: "width 0.25s",
  },
  fieldActive: {
    width: theme.spacing(30),
  },
}));

type Props = {
  isOpen?: boolean;
  placeholder?: string;
  onFilter?: (...args: any[]) => any;
};

const ListingSearch: FunctionComponent<Props> = ({
  isOpen: isOpenProp,
  placeholder,
  onFilter,
}) => {
  const searchRef = useRef<HTMLInputElement>();
  const [isOpen, setIsOpen] = useState(isOpenProp);
  const classes = useStyles();

  useEffect(() => {
    if (isOpen) {
      searchRef.current?.focus();
    }
  }, [isOpen]);

  const handleFilter = (event: any) => {
    if (!onFilter) {
      return;
    }

    const { value } = event.target;

    onFilter(isOpen ? value : undefined);
  };

  const handleClick = () => {
    setIsOpen(!isOpen);

    if (onFilter) {
      onFilter(undefined);
    }
  };

  return (
    <div className={classes.root}>
      <TextField
        type="search"
        placeholder={placeholder}
        className={classNames(classes.field, {
          [classes.fieldActive]: isOpen,
        })}
        onChange={handleFilter}
        inputRef={searchRef}
      />

      <Tooltip title="Search">
        <IconButton onClick={handleClick}>
          <SearchIcon />
        </IconButton>
      </Tooltip>
    </div>
  );
};

export default ListingSearch;
