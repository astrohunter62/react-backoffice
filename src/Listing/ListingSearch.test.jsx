import React from 'react'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import { IconButton } from '@material-ui/core'
import ListingSearch from './ListingSearch'

Enzyme.configure({ adapter: new Adapter() })

it('renders correctly', () => {
  const tree = shallow((
    <ListingSearch
      onClick={() => { }}
      onFilter={() => { }}
    />
  ))

  expect(tree).toMatchSnapshot()
})

it('renders correctly when open', () => {
  const tree = shallow((
    <ListingSearch
      open
      onClick={() => { }}
      onFilter={() => { }}
    />
  ))

  expect(tree).toMatchSnapshot()
})

it('opens if click is triggered', () => {
  const listingSearch = shallow((
    <ListingSearch
      open
      onClick={() => { }}
      onFilter={() => { }}
    />
  ))

  expect(listingSearch.instance().state.open).toEqual(false)

  listingSearch.find(IconButton).simulate('click')

  expect(listingSearch.instance().state.open).toEqual(true)
})

it('filters if open and value changes', () => {
  const onFilter = jest.fn()
  const listingSearch = shallow((
    <ListingSearch
      open
      onClick={() => { }}
      onFilter={onFilter}
    />
  ))

  listingSearch.find('input').simulate('change', { value: 'a' })

  expect(onFilter).toHaveBeenCalled()
})

it('does not filter if not open and value changes', () => {
  const onFilter = jest.fn()
  const listingSearch = shallow((
    <ListingSearch
      open
      onClick={() => { }}
      onFilter={onFilter}
    />
  ))

  expect(listingSearch.instance().state.open).toEqual(false)
  listingSearch.find(IconButton).simulate('click')

  expect(listingSearch.instance().state.open).toEqual(true)

  listingSearch.find('input').simulate('change', { value: 'a' })

  expect(onFilter).toHaveBeenCalled()
})

it('focuses search field on click', () => {
  const mockObject = {
    focus: jest.fn(),
  }

  const listingSearch = shallow((
    <ListingSearch
      onClick={() => { }}
      onFilter={() => {}}
    />
  ))

  listingSearch.instance().searchRef = mockObject

  listingSearch.find(IconButton).simulate('click')

  expect(listingSearch.state().open).toBe(true)

  expect(mockObject.focus).toHaveBeenCalled()
})
