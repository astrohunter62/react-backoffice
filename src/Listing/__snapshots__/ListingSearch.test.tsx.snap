// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<WithStyles(ListingSearchBranch)
  getSearchRef={[Function]}
  onClick={[Function]}
  onFilter={[Function]}
  open={false}
/>
`;

exports[`renders correctly when open 1`] = `
<WithStyles(ListingSearchBranch)
  getSearchRef={[Function]}
  onClick={[Function]}
  onFilter={[Function]}
  open={false}
/>
`;
